src/trace/implementations/ord.rs

```diff
@@ -291,7 +286,9 @@ where
 
 		// if we are supplied a frontier, we should compact.
 		if let Some(frontier) = frontier.as_ref() {
-			OrdValBatch::advance_builder_from(&mut self.result, frontier, initial_key_pos)
+			OrdValBatch::advance_builder_from(&mut self.result, frontier, initial_key_pos);
+			let new_since = antichain_join(self.description.since(), &frontier[..]);
+			self.description = Description::new(self.description.lower(), self.description.upper(), new_since.elements());
```

Comment:
I *think* this is redundant with the changes you've made just above, around lines 229. The `self.description` and `frontier` are going to be static across all of these calls, so that determination at the beginning of the merge set-up should hold for all of the calls as well. But, I could be wrong! Did you add this out of caution, or because there is a specific code path that requires it?

Reply:
Thinking about it now, the fact that `frontier` is static is only because of how the code is used, and it is definitely "more safe" to re-check each time things get called. It is a bit too bad, because 1. we "know" that it won't be a problem, and 2. it is some non-trivial effort to re-determine the since frontier. That's a totally reasonable reason to have this code, though, because in the future someone might mis-use these functions and call with varying frontiers.

Reply:
One option: we could make `frontier` into a `bool` (weird) and use `self.description.since` for the content of `frontier` when the bool is set. I think in the current code we always set `frontier`, but the flexibility exists to merge without compacting. However, I'm not sure there is any value in compacting to anything less than the since frontier of the resulting batch. Maybe we should make that mandatory?

Reply:
(( If we like that idea, ripping out the `frontier` stuff is something either of us can do in a follow-on PR; don't sweat that at the moment, but I am interested to hear your opinion. ))

Reply:
https://github.com/TimelyDataflow/differential-dataflow/blob/ff2aeff062078f0a437f5bcafe3bf8f61bb6fb13/src/trace/implementations/spine_fueled_neu.rs#L855

If I'm not mistaken, this is the only occurrence in which the `begin_merge` function is used.

The "constructor" of the `Merger` takes only the two batches, and the resulting `since` frontier is computed using the `since` frontiers of the two batches. The compaction frontier, however, is not passed to the constructor, but instead is stored inside the `MergeVariant::InProgress` enum to be then passed to the `work` function.

Since the compaction frontier might be greater than the `since` frontier computed in the constructor, we need to update it.

As you correctly pointed out, given how the `work` function is called, the compaction frontier is fixed. 

I think the following refactoring would make it:

* pass the (optional) compaction frontier to the constructor (so we can determine the final `since` frontier right away), rather than to the `work` function.

* if a compaction frontier was passed, the `Merger` sets an internal flag to tell whether or not it should compact when the `work` function is called (the compaction would be performed using the previously computed `since` frontier).

* `MergeVariant::InProgress` variant would then need to store only the two batches being merged and the merger object (i.e. we don't store the compaction frontier anymore).

Reply:
I think this is right, with the caveat that the third bullet probably wants to be "only the two batches being merged and the merger object and the internal flag to tell whether or not it should compact". 

Actually, I'm not certain why the current `frontier: Option<Vec<T>>` isn't part of the merger ... which would also address the problem (prevent using different frontiers as the merge proceeds). But I like you proposal most! We should use the `since` frontier, and have a flag to indicate if compaction should happen (e.g. in case we know that the frontier is below the lower bounds, and would have no effect).

Reply:
I think it's better to store the flag inside the merger, and we set the flag to true if we were provided a compaction frontier on construction. I added a commit with the refactoring.

Reply:
Aaaa, sorry. I didn't notice the last sentence.